src/pydvl/utils/parallel/futures/ray.py

```diff
@@ -0,0 +1,324 @@
+import logging
+import queue
+import sys
+import threading
+import time
+import types
+from concurrent.futures import BrokenExecutor, Executor, Future
+from dataclasses import asdict
+from typing import Any, Callable, Optional, TypeVar
+from weakref import ref
+
+import ray
+
+from pydvl.utils import ParallelConfig
+
+__all__ = ["RayExecutor"]
+
+T = TypeVar("T")
+
+logger = logging.getLogger(__name__)
+
+
+class RayExecutor(Executor):
+    """Asynchronous executor using Ray that implements the concurrent.futures API.
+
+    :param max_workers: Maximum number of concurrent tasks.
+    :param config: instance of :class:`~pydvl.utils.config.ParallelConfig` with cluster address, number of cpus, etc.
+
+    Example:
+        >>> from pydvl.utils.parallel.futures import RayExecutor
+        >>> with RayExecutor() as executor:
+        ...     future = executor.submit(lambda x: x + 1, 1)
+        ...     result = future.result()
+        ...
+        >>> print(result)
+        2
+
+        >>> from pydvl.utils.parallel.futures import RayExecutor
+        >>> with RayExecutor() as executor:
+        ...     results = list(executor.map(lambda x: x + 1, range(5)))
+        ...
+        >>> print(results)
+        [1, 2, 3, 4, 5]
+    """
+
+    def __init__(
+        self,
+        max_workers: Optional[int] = None,
+        *,
+        config: ParallelConfig = ParallelConfig(),
+    ):
+        if config.backend != "ray":
+            raise ValueError(
+                f"Parallel backend must be set to 'ray' and not {config.backend}"
+            )
+        if max_workers is not None:
+            if max_workers <= 0:
+                raise ValueError("max_workers must be greater than 0")
+            max_workers = max_workers
+
+        config_dict = asdict(config)
+        config_dict.pop("backend")
+        config_dict.pop("n_workers")
+        if "address" not in config_dict:
+            config_dict["num_cpus"] = max_workers
+        self.config = config_dict
+        if not ray.is_initialized():
+            ray.init(**self.config)
+
+        self._max_workers = max_workers
+        if self._max_workers is None:
+            self._max_workers = int(ray._private.state.cluster_resources()["CPU"])
```

Comment:
If we stick to 1 CPU per job, then:

```suggestion
        self.n_cpus_per_job = 1  # TODO: make configurable
        available_cpus = int(ray._private.state.cluster_resources()["CPU"])
        if max_workers is None:
            max_workers = available_cpus // self.n_cpus_per_job
        elif max_workers*self.n_cpus_per_job > available_cpus:
            logger.warning(f"max_workers*n_cpus_per_job cannot be larger than the available cluster resources ({available_cpus} CPUs)")
            max_workers = available_cpus // self.n_cpus_per_job
        self._max_workers = max_workers
```

And we should probably allow configuration of `n_cpus_per_job` in the config.

Reply:
I added this now as a field in ParallelConfig.